packetbeat/protos/dns/dns.go

```diff
@@ -419,167 +454,249 @@ func addDnsToMapStr(m common.MapStr, dns *mkdns.Msg, authority bool, additional
 	mapStrArray := make([]common.MapStr, len(records))
 	for i, rr := range records {
 		rrHeader := rr.Header()
-		rrType := rrHeader.Rrtype
 
-		mapStr := common.MapStr{
-			"name":  rrHeader.Name,
-			"type":  dnsTypeToString(rrType),
-			"class": dnsClassToString(rrHeader.Class),
-			"ttl":   strconv.FormatInt(int64(rrHeader.Ttl), 10),
+		mapStr := rrToMapStr(rr)
+		if len(mapStr) == 0 { // OPT pseudo-RR returns an empty MapStr
+			resizeStrArray := make([]common.MapStr, len(mapStrArray)-1)
+			copy(resizeStrArray, mapStrArray)
+			mapStrArray = resizeStrArray
+			continue
 		}
+		mapStr["name"] = rrHeader.Name
+		mapStr["type"] = dnsTypeToString(rrHeader.Rrtype)
+		mapStr["class"] = dnsClassToString(rrHeader.Class)
+		mapStr["ttl"] = strconv.FormatInt(int64(rrHeader.Ttl), 10)
 		mapStrArray[i] = mapStr
-
-		switch x := rr.(type) {
-		default:
-			// We don't have special handling for this type
-			logp.Debug("dns", "No special handling for RR type %s", dnsTypeToString(rrType))
-			unsupportedRR := new(mkdns.RFC3597)
-			err := unsupportedRR.ToRFC3597(x)
-			if err == nil {
-				rData, err := hexStringToString(unsupportedRR.Rdata)
-				mapStr["data"] = rData
-				if err != nil {
-					logp.Debug("dns", "%s", err.Error())
-				}
-			} else {
-				logp.Debug("dns", "Rdata for the unhandled RR type %s could not be fetched", dnsTypeToString(rrType))
-			}
-		case *mkdns.A:
-			mapStr["data"] = x.A.String()
-		case *mkdns.AAAA:
-			mapStr["data"] = x.AAAA.String()
-		case *mkdns.CNAME:
-			mapStr["data"] = x.Target
-		case *mkdns.MX:
-			mapStr["preference"] = x.Preference
-			mapStr["data"] = x.Mx
-		case *mkdns.NS:
-			mapStr["data"] = x.Ns
-		case *mkdns.PTR:
-			mapStr["data"] = x.Ptr
-		case *mkdns.RFC3597:
-			// Miekg/dns lib doesn't handle this type
-			//TODO: write a test for this.
-			logp.Debug("dns", "Unknown RR type %s", dnsTypeToString(rrType))
-			rData, err := hexStringToString(x.Rdata)
-			mapStr["data"] = rData
-			if err != nil {
-				logp.Debug("dns", "%s", err.Error())
-			}
-		case *mkdns.SOA:
-			mapStr["rname"] = x.Mbox
-			mapStr["serial"] = x.Serial
-			mapStr["refresh"] = x.Refresh
-			mapStr["retry"] = x.Retry
-			mapStr["expire"] = x.Expire
-			mapStr["minimum"] = x.Minttl
-			mapStr["data"] = x.Ns
-		case *mkdns.SRV:
-			mapStr["priority"] = x.Priority
-			mapStr["weight"] = x.Weight
-			mapStr["port"] = x.Port
-			mapStr["data"] = x.Target
-		case *mkdns.TXT:
-			mapStr["data"] = strings.Join(x.Txt, " ")
-		}
 	}
-
 	return mapStrArray
 }
 
-// dnsQuestionToString converts a Question to a string.
-func dnsQuestionToString(q mkdns.Question) string {
-	name := q.Name
+// Convert all RDATA fields of a RR to a single string
+// fields are ordered alphabetically with 'data' as the last element
+//
+// TODO An improvement would be to replace 'data' by the real field name
+// It would require some changes in unit tests
+func rrToString(rr mkdns.RR) string {
+	var st string
+	var keys []string
```

Comment:
A small optimization would be to allocate the slice once since we know the number of keys. `keys = make([]string, 0, len(mapStr))`